test/jdk/sun/net/www/protocol/file/NonLocalFtpFallback.java

```diff
@@ -0,0 +1,141 @@
+import java.util.HashSet;
+import java.util.Random;
+import java.util.Set;
+
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * @test
+ * @bug 8353662
+ * @summary Verify long-standing behavior of resolving non-local file URLs using FTP.
+ * @run junit NonLocalFtpFallback
+ */
+
+public class NonLocalFtpFallback {
+
+    // Port 21 may not be available, use an HTTP proxy with an ephemeral port
+    private HttpServer proxyServer;
+
+    // The file requested in this test
+    private Path file;
+
+    // FTP URIs requested by the proxy client
+    private Set<URI> uris = new HashSet<>();
+
+    /**
+     * Set up the HTTP proxy used for serving FTP in this test
+     *
+     * @throws IOException if an unexpected IO error occurs
+     */
+    @BeforeEach
+    public void setup() throws IOException {
+        // Create a file with some random data
+        byte[] data = new byte[512];
+        new Random().nextBytes(data);
+        file = Files.write(Path.of("ftp-file.txt"), data);
+
+        // Set up an HTTP proxy server
+        proxyServer = HttpServer.create();
+        // Bind to the loopback address with an ephemeral port
+        InetAddress loopbackAddress = InetAddress.getLoopbackAddress();
+        proxyServer.bind(new InetSocketAddress(loopbackAddress, 0), 0);
+        // Handler for the FTP proxy request
+        proxyServer.createContext("/", new HttpHandler() {
+            @Override
+            public void handle(HttpExchange exchange) throws IOException {
+                // Record the URI requested
+                uris.add(exchange.getRequestURI());
+                // Send the data
+                exchange.sendResponseHeaders(200, Files.size(file));
+                try (OutputStream out = exchange.getResponseBody()) {
+                    Files.copy(file, out);
+                }
+                // Complete the exchange
+                exchange.close();
+            }
+        });
+        // Start the proxy server
+        proxyServer.start();
+    }
+
+    /**
+     * Shut down proxy server and clean up files created
+     *
+     * @throws IOException if an unexpected IO error occurs
+     */
+    @AfterEach
+    public void destroy() throws IOException {
+        proxyServer.stop(2);
+        Files.delete(file);
+    }
+
+    /**
+     * Verifies the long-standing and unspecified FTP fallback feature where the file
+     * URL scheme handler attempts an FTP connection for non-local files.
+     *
+     * The non-local file URL used here is of the form file://127.0.0.1/path. Since the
+     * host component here is not equal to "localhost", this is considered a non-local
+     * URL.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void verifyNonLocalFtpFallback() throws Exception {
+        URL localURL = file.toUri().toURL();
+        URL nonLocalURL = new URL("file", "127.0.0.1", localURL.getFile());
```

Comment:
The URL should not assume that the loopback is an IPv4 address.

Reply:
> The URL should not assume that the loopback is an IPv4 address. Or is this handled by the proxy?

Right. The following returns "127.0.0.1" for me, would that be the correct API to use?

`InetAddress.getLoopbackAddress().getHostAddress()`

Reply:
Ok - I see that 127.0.0.1 should not be directly used. Would it work if we replaced that with a fake host name then?

Reply:
> Ok - I see that 127.0.0.1 should not be directly used. Would it work if we replaced that with a fake host name then?

Yeah, you're right, the FTP request never materializes (our proxy just swallows it). So we only use it to assert against what was requested from the proxy, we can use a fake host name.

Any preference/precedence for fake host names?

Reply:
Generally looks good. "127.0.0.1" rings alarm bells because we're trying to make our tests independent of the IP versions supported by the machine they run on.

Reply:
> Generally looks good.

Does this mean you think the PR is good for approval now that the IP version aspect has been taken care of?

Or would you like more time/people to review this test-only enhancement?

Reply:
Sorry - thanks for pinging me.
I tested the change in our CI and the new test seem stable. Approved.